File: sdks/typescript/src/cli-folder.ts
/** Generate and maintain the .cli/ reference folder per ACLI spec §1.3. */

import * as fs from 'node:fs';
import * as path from 'node:path';
import type { CommandTree } from './introspect.js';

/** Generate the .cli/ folder with all required files. */
export function generateCliFolder(
  tree: CommandTree,
  targetDir?: string,
): string {
  const root = path.join(targetDir ?? process.cwd(), '.cli');
  fs.mkdirSync(path.join(root, 'examples'), { recursive: true });
  fs.mkdirSync(path.join(root, 'schemas'), { recursive: true });

  // commands.json
  fs.writeFileSync(
    path.join(root, 'commands.json'),
    JSON.stringify(tree, null, 2) + '\n',
  );

  // README.md
  writeReadme(root, tree);

  // Example scripts
  writeExamples(root, tree);

  // changelog.md (create if missing)
  const changelog = path.join(root, 'changelog.md');
  if (!fs.existsSync(changelog)) {
    fs.writeFileSync(
      changelog,
      `# Changelog\n\n## ${tree.version}\n\n- Initial release\n`,
    );
  }

  return root;
}

/** Check whether .cli/commands.json is out of date. */
export function needsUpdate(tree: CommandTree, targetDir?: string): boolean {
  const root = path.join(targetDir ?? process.cwd(), '.cli');
  const commandsFile = path.join(root, 'commands.json');
  if (!fs.existsSync(commandsFile)) return true;
  try {
    const existing = JSON.parse(fs.readFileSync(commandsFile, 'utf-8'));
    return JSON.stringify(existing) !== JSON.stringify(tree);
  } catch {
    return true;
  }
}

function writeReadme(cliDir: string, tree: CommandTree): void {
  const lines = [
    `# ${tree.name}`,
    '',
    `Version: ${tree.version}`,
    `ACLI version: ${tree.acli_version}`,
    '',
    '## Commands',
    '',
  ];
  for (const cmd of tree.commands) {
    lines.push(`### ${cmd.name}`, '', cmd.description, '');
    if (cmd.idempotent !== undefined) {
      lines.push(`Idempotent: ${cmd.idempotent}`, '');
    }
  }
  fs.writeFileSync(path.join(cliDir, 'README.md'), lines.join('\n') + '\n');
}

function writeExamples(cliDir: string, tree: CommandTree): void {
  for (const cmd of tree.commands) {
    if (!cmd.examples?.length) continue;
    const lines = [
      '#!/usr/bin/env bash',
      `# Examples for: ${cmd.name}`,
      '',
    ];
    for (const ex of cmd.examples) {
      lines.push(`# ${ex.description}`, ex.invocation, '');
    }
    fs.writeFileSync(
      path.join(cliDir, 'examples', `${cmd.name}.sh`),
      lines.join('\n'),
    );
  }
}

File: sdks/go/cli_folder.go
package acli

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// GenerateCliFolder writes .cli/ per spec §1.3.
func GenerateCliFolder(tree *CommandTree, targetDir string) (string, error) {
	root := filepath.Join(targetDir, ".cli")
	if targetDir == "" {
		root = ".cli"
	}
	if err := os.MkdirAll(filepath.Join(root, "examples"), 0o755); err != nil {
		return "", err
	}
	if err := os.MkdirAll(filepath.Join(root, "schemas"), 0o755); err != nil {
		return "", err
	}

	b, err := json.MarshalIndent(tree, "", "  ")
	if err != nil {
		return "", err
	}
	if err := os.WriteFile(filepath.Join(root, "commands.json"), append(b, '\n'), 0o644); err != nil {
		return "", err
	}
	if err := writeReadme(root, tree); err != nil {
		return "", err
	}
	if err := writeExamples(root, tree); err != nil {
		return "", err
	}
	changelog := filepath.Join(root, "changelog.md")
	if _, err := os.Stat(changelog); os.IsNotExist(err) {
		body := fmt.Sprintf("# Changelog\n\n## %s\n\n- Initial release\n", tree.Version)
		if err := os.WriteFile(changelog, []byte(body), 0o644); err != nil {
			return "", err
		}
	}
	return root, nil
}

func writeReadme(cliDir string, tree *CommandTree) error {
	var lines []string
	lines = append(lines, "# "+tree.Name, "")
	lines = append(lines, "Version: "+tree.Version)
	lines = append(lines, "ACLI version: "+tree.AcliVersion, "")
	lines = append(lines, "## Commands", "")
	for _, cmd := range tree.Commands {
		lines = append(lines, "### "+cmd.Name, "", cmd.Description, "")
		if cmd.Idempotent != nil {
			lines = append(lines, fmt.Sprintf("Idempotent: %v", cmd.Idempotent), "")
		}
	}
	return os.WriteFile(filepath.Join(cliDir, "README.md"), []byte(joinLines(lines)), 0o644)
}

func writeExamples(cliDir string, tree *CommandTree) error {
	for _, cmd := range tree.Commands {
		if len(cmd.Examples) == 0 {
			continue
		}
		var lines []string
		lines = append(lines, "#!/usr/bin/env bash", fmt.Sprintf("# Examples for: %s", cmd.Name), "")
		for _, ex := range cmd.Examples {
			lines = append(lines, "# "+ex.Description, ex.Invocation, "")
		}
		p := filepath.Join(cliDir, "examples", cmd.Name+".sh")
		if err := os.WriteFile(p, []byte(joinLines(lines)), 0o644); err != nil {
			return err
		}
	}
	return nil
}

func joinLines(lines []string) string {
	out := ""
	for i, l := range lines {
		if i > 0 {
			out += "\n"
		}
		out += l
	}
	return out + "\n"
}

// NeedsUpdate returns true if .cli/commands.json is missing or stale.
func NeedsUpdate(tree *CommandTree, targetDir string) bool {
	root := filepath.Join(targetDir, ".cli")
	if targetDir == "" {
		root = ".cli"
	}
	p := filepath.Join(root, "commands.json")
	data, err := os.ReadFile(p)
	if err != nil {
		return true
	}
	var existing map[string]any
	if err := json.Unmarshal(data, &existing); err != nil {
		return true
	}
	cur, err := json.Marshal(tree)
	if err != nil {
		return true
	}
	var curMap map[string]any
	if err := json.Unmarshal(cur, &curMap); err != nil {
		return true
	}
	b1, _ := json.Marshal(existing)
	b2, _ := json.Marshal(curMap)
	return string(b1) != string(b2)
}
